patchwork-fml/src/main/java/net/minecraftforge/fml/ModList.java

```diff
@@ -53,73 +53,69 @@ public boolean isLoaded(String modId) {
 	}
 
-	//return null if it does not have annotation data
-	private static String getAnnotationHolderModid(String modid) {
-		ModContainer modContainer = FabricLoader.getInstance().getModContainer(modid).orElse(null);
+	private ModFileInfo getModFileByContainer(ModContainer modContainer) {
+		return modFileInfoMap.computeIfAbsent(modContainer, this::createModFileInfo);
+	}
 
-		if (modContainer == null) {
-			LOGGER.error("Trying to access annotation data of a missing mod " + modid);
-			LOGGER.catching(new Throwable());
-			return null;
+	private ModFileInfo createModFileInfo(ModContainer modContainer) {
+		ModMetadata metadata = modContainer.getMetadata();
+
+		// First try to find a patchwork:annotations entry for this file. If it exists, then this is the "primary" mod
+		// for a given JAR file.
+		CustomValue annotations = metadata.getCustomValue("patchwork:annotations");
+
+		if (annotations != null) {
+			String annotationJsonLocation = annotations.getAsString();
+
+			return new ModFileInfo(modContainer, annotationJsonLocation);
 		}
 
-		if (!isPatchedMod(modContainer)) {
-			return null;
+		// If there is no annotation data indicative of a primary mod file, try to then find the parent (primary) mod ID.
+		// This indicates that this is a dummy JiJ mod created by Patchwork Patcher.
+		CustomValue parent = metadata.getCustomValue("patchwork:parent");
+
+		if (parent != null) {
+			return getModFileById(parent.getAsString());
 		}
 
-		CustomValue parent = modContainer.getMetadata().getCustomValue("patchwork:parent");
+		// This mod lacks annotation data or a parent mod ID.
+		// Check to see if it was run through an old version of Patcher (if it lacks both the parent and annotations
+		// attributes but has the source attribute)
+		CustomValue source = modContainer.getMetadata().getCustomValue("patchwork:source");
 
-		if (parent == null) {
-			//it's not a jij mod
-			return modid;
+		if (source != null) {
+			CustomValue.CvObject object = source.getAsObject();
+			String loader = object.get("loader").getAsString();
+
+			if (loader.equals("forge")) {
+				LOGGER.warn("A mod was patched with an old version of Patchwork Patcher, please re-patch it! "
+						+ "No annotation data is available for " + metadata.getId() + " (loaded from " + modContainer.getRootPath() + ")");
+			}
 		}
 
-		return parent.getAsString();
+		// Either a patchwork mod missing annotation data, or a normal Fabric mod.
+		return new ModFileInfo();
 	}
 
-	public static boolean isPatchedMod(ModContainer modContainer) {
-		CustomValue source = modContainer.getMetadata().getCustomValue("patchwork:source");
-
-		if (source == null) {
-			return false;
+	public List<ModFileScanData> getAllScanData() {
+		if (allScanDataCache == null) {
+			allScanDataCache = FabricLoader.getInstance().getAllMods()
+					.stream()
+					.map(modContainer -> modContainer.getMetadata().getId())
+					.map(modid -> getModFileById(modid).getFile().getScanResult())
+					.distinct()
```

Comment:
ModFileScanData doesn't implement hashcode or equals so this does nothing.

Reply:
In this case there is only one instance of ModFileScanData for a given mod file (mod files can be shared by multiple mod containers), therefore comparison by object identity alone (`==`) is sufficient. I'll add a comment for this.